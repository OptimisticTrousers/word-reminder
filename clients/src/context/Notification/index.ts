export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
