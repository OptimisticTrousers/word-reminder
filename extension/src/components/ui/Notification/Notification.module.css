.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
