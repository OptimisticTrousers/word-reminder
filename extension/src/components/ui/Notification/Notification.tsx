import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
