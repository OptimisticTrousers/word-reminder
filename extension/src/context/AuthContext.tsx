import { createContext, FC, useReducer } from "react";

interface IUser {
  _id: string;
  username: string;
  theme: string;
  words: IUserWord[];
}

interface IWord {
  _id: string;
  word: string;
  origin: string;
  phonetic: string;
  audio: string;
  meanings: [
    {
      partOfSpeech: string;
      definitions: [
        {
          definition: string;
          example: string;
          synonyms: string[];
          antonyms: string[];
        }
      ];
    }
  ];
}

export interface IUserWord {
  _id: string;
  userId: IUser | string;
  word: IWord;
  learned: boolean;
}

interface AuthState {
  user: IUser | null;
}

interface Action {
  type: string;
  payload: IUser;
}

interface AuthContext {
  state: AuthState;
  dispatch: (action: Action) => void;
}

const authReducer = (state: AuthState, action: Action) => {
  switch (action.type) {
    // Login action will always contain a payload of user object
    case "LOGIN":
      return { user: action.payload };
    // Remove user object on logout action dispatch
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContext = createContext<AuthContext>({} as AuthContext);

interface Props {
  children: JSX.Element[] | JSX.Element;
}

const AuthProvider: FC<Props> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
