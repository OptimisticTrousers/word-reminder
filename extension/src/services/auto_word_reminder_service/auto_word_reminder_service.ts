import { AutoWordReminderParams } from "common";

import { service } from "../service";

export const autoWordReminderService = (function (service) {
  const { get, post, put, remove, VITE_API_DOMAIN } = service;

  function getAutoWordReminder(userId: string) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: {
        credentials: "include",
      },
    });
  }

  function createAutoWordReminder({
    userId,
    body,
  }: {
    userId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function updateAutoWordReminder({
    userId,
    autoWordReminderId,
    body,
  }: {
    userId: string;
    autoWordReminderId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return put({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function deleteAutoWordReminder({
    userId,
    autoWordReminderId,
  }: {
    userId: string;
    autoWordReminderId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { credentials: "include" },
    });
  }

  return {
    createAutoWordReminder,
    getAutoWordReminder,
    deleteAutoWordReminder,
    updateAutoWordReminder,
  };
})(service);
