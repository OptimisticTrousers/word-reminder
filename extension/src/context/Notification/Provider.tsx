import { useReducer, ReactNode } from "react";

import { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
import { Notification } from "../../components/ui/Notification/Notification";

interface Props {
  children: ReactNode;
}

interface NotificationState {
  isVisible: boolean;
  type: string;
  message: string;
}

interface NotificationAction {
  payload: { message: string };
  type: string;
}

export interface NotificationContext {
  showNotification: (type: string, message: string) => void;
  dismissNotification: () => void;
}

const notificationReducer = (
  state: NotificationState,
  action: NotificationAction
) => {
  switch (action.type) {
    case NOTIFICATION_ACTIONS.SUCCESS:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.SUCCESS,
        message: action.payload.message,
      };
    case NOTIFICATION_ACTIONS.ERROR:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.ERROR,
        message: action.payload.message,
      };
    case DISAPPEAR:
      return { ...state, isVisible: false, type: "", message: "" };
    default: {
      throw Error("Unknown action: " + action.type);
    }
  }
};

// Provide ability to spawn Toast notification globally within the application
export function NotificationProvider({ children }: Props) {
  const [state, dispatch] = useReducer(notificationReducer, {
    isVisible: false,
    type: "",
    message: "",
  });

  function dismissNotification() {
    dispatch({
      type: DISAPPEAR,
      payload: { message: "" },
    });
  }

  function showNotification(type: string, message: string): void {
    const delay = 5000;

    dispatch({ type, payload: { message } });

    setTimeout(() => {
      dismissNotification();
    }, delay);
  }

  const { isVisible, type, message } = state;

  return (
    <NotificationContext.Provider
      value={{ showNotification, dismissNotification }}
    >
      {isVisible && <Notification type={type} message={message} />}
      {children}
    </NotificationContext.Provider>
  );
}

const DISAPPEAR = "disappear";
