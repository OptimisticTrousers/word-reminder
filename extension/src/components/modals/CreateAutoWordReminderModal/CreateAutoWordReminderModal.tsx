import { SortMode, User } from "common";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { ToggleModal } from "../types";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { ChangeEvent, useContext, useState } from "react";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { ModalContainer } from "../ModalContainer";
import styles from "./CreateAutoWordReminderModal.module.css";
import { Duration } from "../../ui/Duration";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";
import { Reminder } from "../../ui/Reminder";
import { unitsToMs } from "../../../utils/date/date";

interface Props {
  toggleModal: ToggleModal;
}

export const AutoCreateWordReminderModal = CSSModules(
  function ({ toggleModal }: Props) {
    const [reminder, setReminder] = useState("");
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const queryClient = useQueryClient();
    const { isPending, mutate } = useMutation({
      mutationFn: autoWordReminderService.createAutoWordReminder,
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          CREATE_AUTO_WORD_REMINDER_NOTIFICATION_MSGS.createAutoWordReminder()
        );
        queryClient.invalidateQueries({
          queryKey: ["autoWordReminders"],
          exact: true,
        });
      },
      onError: showNotificationError,
      onSettled: () => {
        toggleModal();
      },
    });

    function handleReminderChange(event: ChangeEvent<HTMLInputElement>) {
      setReminder(event.target.value);
    }

    function handleCreate(formData: FormData) {
      const time = {
        minutes: Number(formData.get("minutes") as string),
        hours: Number(formData.get("hours") as string),
        days: Number(formData.get("days") as string),
        weeks: Number(formData.get("weeks") as string),
      };
      mutate({
        userId,
        body: {
          reminder: formData.get("reminder") as string,
          create_now: Boolean(formData.get("create_now") as string),
          duration: unitsToMs(time),
          word_count: Number(formData.get("word_count") as string),
          is_active: Boolean(formData.get("is_active") as string),
          has_reminder_onload: Boolean(
            formData.get("has_reminder_onload") as string
          ),
          has_learned_words: Boolean(
            formData.get("has_learned_words") as string
          ),
          sort_mode: formData.get("sort_mode") as SortMode,
        },
      });
    }

    return (
      <ModalContainer
        title="Create Auto Word Reminder"
        toggleModal={toggleModal}
      >
        <form styleName="modal__form" action={handleCreate}>
          <Reminder
            disabled={false}
            value={reminder}
            handleChange={handleReminderChange}
          />
          <Duration
            disabled={false}
            defaultValues={{ minutes: 0, hours: 0, days: 0, weeks: 0 }}
          />
          <div styleName="modal__control">
            <label styleName="modal__label" htmlFor="word_count">
              Word Count
            </label>
            <input
              styleName="modal__input"
              type="number"
              id="word_count"
              name="word_count"
              required
              max={99}
            />
          </div>
          <fieldset styleName="modal__fieldset">
            <legend styleName="modal__legend">Options</legend>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="create_now">
                Create Now
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="create_now"
                name="create_now"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="is_active">
                Is Active
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="is_active"
                name="is_active"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_reminder_onload">
                Has Reminder Onload
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_reminder_onload"
                name="has_reminder_onload"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_learned_words">
                Has Learned Words
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_learned_words"
                name="has_learned_words"
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="sort_mode">
                Sort Mode
              </label>
              <select
                styleName="modal__select"
                name="sort_mode"
                id="sort_mode"
                required
                defaultValue={SortMode.Random}
              >
                <option styleName="modal__option" value={SortMode.Newest}>
                  Newest
                </option>
                <option styleName="modal__option" value={SortMode.Oldest}>
                  Oldest
                </option>
                <option styleName="modal__option" value={SortMode.Random}>
                  Random
                </option>
              </select>
            </div>
          </fieldset>
          <button styleName="modal__button" type="submit" disabled={isPending}>
            Create
          </button>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const CREATE_AUTO_WORD_REMINDER_NOTIFICATION_MSGS = {
  createAutoWordReminder: () => {
    return "Word reminders will now be created automatically!";
  },
};
