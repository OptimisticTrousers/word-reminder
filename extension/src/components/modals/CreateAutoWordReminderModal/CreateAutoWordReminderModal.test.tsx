import { render, screen } from "@testing-library/react";
import { createRoutesStub, Outlet } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { AutoCreateWordReminderModal } from "./CreateAutoWordReminderModal";
import userEvent from "@testing-library/user-event";
import { Order } from "common";
import { NotificationProvider } from "../../../context/Notification";
import { Mock } from "vitest";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";

vi.mock("../ModalContainer/ModalContainer");

describe("CreateAutoWordReminderModal component", () => {
  const testUser = {
    id: "1",
  };

  function setup({
    toggleModal,
    queryClient,
  }: {
    toggleModal: Mock;
    queryClient: QueryClient;
  }) {
    const searchParams = new URLSearchParams();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <NotificationProvider>
                  <QueryClientProvider client={queryClient}>
                    <AutoCreateWordReminderModal
                      searchParams={searchParams}
                      toggleModal={toggleModal}
                    />
                  </QueryClientProvider>
                </NotificationProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  const autoWordReminder = {
    id: "1",
    reminder: {
      minutes: 30,
      hours: 1,
      days: 7,
      weeks: 1,
      months: 1,
    },
    create_now: false,
    duration: {
      minutes: 2,
      hours: 2,
      days: 1,
      weeks: 3,
      months: 2,
    },
    word_count: 7,
    is_active: false,
    has_reminder_onload: false,
    has_learned_words: true,
    order: Order.Oldest,
  };

  const status = 200;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders form with all inputs", async () => {
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const { asFragment } = setup({ toggleModal: mockToggleModal, queryClient });

    const reminder = screen.getByText("Reminder");
    const duration = screen.getByText("Duration");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const order = screen.getByLabelText("Order", { selector: "select" });
    const notification = screen.queryByRole("dialog");
    const modalHeading = screen.getByTestId("modal-heading");
    expect(modalHeading).toBeInTheDocument();
    expect(modalHeading).toHaveTextContent(
      "Automatically Create Word Reminder"
    );
    expect(notification).not.toBeInTheDocument();
    expect(reminder).toBeInTheDocument();
    expect(duration).toBeInTheDocument();
    expect(wordCount).toBeInTheDocument();
    expect(createNow).toBeInTheDocument();
    expect(createNow).toBeChecked();
    expect(isActive).toBeInTheDocument();
    expect(isActive).toBeChecked();
    expect(hasReminderOnload).toBeInTheDocument();
    expect(hasReminderOnload).toBeChecked();
    expect(hasLearnedWords).toBeInTheDocument();
    expect(hasLearnedWords).not.toBeChecked();
    expect(order).toBeInTheDocument();
    expect(order).toHaveValue(Order.Random);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the functions to automatically create an auto word reminder", async () => {
    const mockAutoCreateWordReminder = vi
      .spyOn(autoWordReminderService, "createAutoWordReminder")
      .mockImplementation(async () => {
        return {
          json: { autoWordReminder },
          status,
        };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const [reminderMinutes, durationMinutes] =
      screen.getAllByLabelText("Minutes");
    const [reminderHours, durationHours] = screen.getAllByLabelText("Hours");
    const [reminderDays, durationDays] = screen.getAllByLabelText("Days");
    const [reminderWeeks, durationWeeks] = screen.getAllByLabelText("Weeks");
    const [reminderMonths, durationMonths] = screen.getAllByLabelText("Months");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });
    const createNow = screen.getByLabelText("Create Now", {
      selector: "input",
    });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const hasLearnedWords = screen.getByLabelText("Has Learned Words", {
      selector: "input",
    });
    const order = screen.getByLabelText("Order", { selector: "select" });

    await user.type(reminderMinutes, "30");
    await user.type(reminderHours, "1");
    await user.type(reminderDays, "7");
    await user.type(reminderWeeks, "1");
    await user.type(reminderMonths, "1");
    await user.type(durationMinutes, "2");
    await user.type(durationHours, "2");
    await user.type(durationDays, "1");
    await user.type(durationWeeks, "3");
    await user.type(durationMonths, "2");
    await user.type(wordCount, "7");
    await user.click(createNow);
    await user.click(isActive);
    await user.click(hasReminderOnload);
    await user.click(hasLearnedWords);
    await user.selectOptions(order, Order.Oldest);
    await user.click(createButton);

    const notification = screen.getByRole("dialog", {
      name: "Word reminders will now be created automatically!",
    });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
    expect(mockInvalidateQueries).toHaveBeenCalledWith({
      queryKey: ["wordReminders", Object.fromEntries(new URLSearchParams())],
      exact: true,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockAutoCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockAutoCreateWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        reminder: {
          minutes: 30,
          hours: 1,
          days: 7,
          weeks: 1,
          months: 1,
        },
        create_now: false,
        duration: {
          minutes: 2,
          hours: 2,
          days: 1,
          weeks: 3,
          months: 2,
        },
        word_count: 7,
        is_active: false,
        has_reminder_onload: false,
        has_learned_words: true,
        order: Order.Oldest,
      },
    });
  });

  it("calls the functions to show a notification error", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockAutoCreateWordReminder = vi
      .spyOn(autoWordReminderService, "createAutoWordReminder")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const [reminderHours] = screen.getAllByLabelText("Hours");
    const [, durationWeeks] = screen.getAllByLabelText("Weeks");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminderHours, "1");
    await user.type(durationWeeks, "2");
    await user.type(wordCount, "7");
    await user.click(createButton);

    const notification = screen.getByRole("dialog", { name: message });
    expect(notification).toBeInTheDocument();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockAutoCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockAutoCreateWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        reminder: {
          minutes: 0,
          hours: 1,
          days: 0,
          weeks: 0,
          months: 0,
        },
        duration: {
          minutes: 0,
          hours: 0,
          days: 0,
          weeks: 2,
          months: 0,
        },
        create_now: true,
        word_count: 7,
        is_active: true,
        has_reminder_onload: true,
        has_learned_words: false,
        order: Order.Random,
      },
    });
  });

  it("disables the create button when the mutation is loading", async () => {
    const delay = 50;
    const mockCreateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "createAutoWordReminder")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { autoWordReminder }, status });
          }, delay);
        });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const [reminderHours] = screen.getAllByLabelText("Hours");
    const [, durationWeeks] = screen.getAllByLabelText("Weeks");
    const wordCount = screen.getByLabelText("Word Count", {
      selector: "input",
    });
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminderHours, "1");
    await user.type(durationWeeks, "2");
    await user.type(wordCount, "7");
    await user.click(createButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(createButton).toBeDisabled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockToggleModal).not.toHaveBeenCalled();
    expect(mockCreateAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateAutoWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      body: {
        reminder: {
          minutes: 0,
          hours: 1,
          days: 0,
          weeks: 0,
          months: 0,
        },
        duration: {
          minutes: 0,
          hours: 0,
          days: 0,
          weeks: 2,
          months: 0,
        },
        word_count: 7,
        is_active: true,
        create_now: true,
        has_reminder_onload: true,
        has_learned_words: false,
        order: Order.Random,
      },
    });
  });

  it("calls the functions to close a modal", async () => {
    const mockCreateAutoWordReminder = vi
      .spyOn(autoWordReminderService, "createAutoWordReminder")
      .mockImplementation(async () => {
        return { json: { autoWordReminder }, status };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient();
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });

    const closeButton = screen.getByRole("button", { name: "Close modal" });
    await user.click(closeButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockCreateAutoWordReminder).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  describe("form validation", () => {
    it("only allows the user to enter up to 99 for wordCount", async () => {
      const mockCreateAutoWordReminder = vi
        .spyOn(autoWordReminderService, "createAutoWordReminder")
        .mockImplementation(async () => {
          return { json: { autoWordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const [reminderHours] = screen.getAllByLabelText("Hours");
      const [, durationWeeks] = screen.getAllByLabelText("Weeks");
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminderHours, "1 hours");
      await user.type(durationWeeks, "2 weeks");
      await user.type(wordCount, "100");
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateAutoWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow the user to submit when wordCount is empty", async () => {
      const mockCreateAutoWordReminder = vi
        .spyOn(autoWordReminderService, "createAutoWordReminder")
        .mockImplementation(async () => {
          return { json: { autoWordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient();
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const [reminderHours] = screen.getAllByLabelText("Hours");
      const [, durationWeeks] = screen.getAllByLabelText("Weeks");
      const wordCount = screen.getByLabelText("Word Count", {
        selector: "input",
      });
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminderHours, "1 hour");
      await user.type(durationWeeks, "2 weeks");
      await user.clear(wordCount);
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockCreateAutoWordReminder).not.toHaveBeenCalled();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });
  });
});
