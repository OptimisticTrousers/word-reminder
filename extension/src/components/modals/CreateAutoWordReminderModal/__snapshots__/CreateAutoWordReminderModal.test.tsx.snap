// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateAutoWordReminderModal component > renders form with all inputs 1`] = `
<DocumentFragment>
  <div>
    <h2
      data-testid="modal-heading"
    >
      Create Auto Word Reminder
    </h2>
    <button
      aria-label="Close modal"
    >
      X
    </button>
    <div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal__form_42f588"
      >
        <div
          class="_reminder_a2fc7e"
        >
          <p
            class="_reminder__description_a2fc7e"
          />
          <label
            class="_reminder__label_a2fc7e"
          >
            Reminder
            <input
              class="_reminder__input_a2fc7e"
              name="reminder"
              placeholder="Enter a valid cron expression (ie. * * * * * or */5 * * * *)"
              required=""
              type="text"
              value=""
            />
          </label>
        </div>
        <fieldset
          class="_duration_03697c"
        >
          <legend
            class="_duration__legend_03697c"
          >
            Duration
          </legend>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="minutes"
            >
              Minutes
            </label>
            <input
              class="_duration__input_03697c"
              id="minutes"
              max="59"
              min="0"
              name="minutes"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="hours"
            >
              Hours
            </label>
            <input
              class="_duration__input_03697c"
              id="hours"
              max="23"
              min="0"
              name="hours"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="days"
            >
              Days
            </label>
            <input
              class="_duration__input_03697c"
              id="days"
              max="30"
              min="0"
              name="days"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="_duration__control_03697c"
          >
            <label
              class="_duration__label_03697c"
              for="weeks"
            >
              Weeks
            </label>
            <input
              class="_duration__input_03697c"
              id="weeks"
              max="51"
              min="0"
              name="weeks"
              required=""
              type="number"
              value="0"
            />
          </div>
        </fieldset>
        <div
          class="_modal__control_42f588"
        >
          <label
            class="_modal__label_42f588"
            for="word_count"
          >
            Word Count
          </label>
          <input
            class="_modal__input_42f588"
            id="word_count"
            max="99"
            name="word_count"
            required=""
            type="number"
          />
        </div>
        <fieldset
          class="_modal__fieldset_42f588"
        >
          <legend
            class="_modal__legend_42f588"
          >
            Options
          </legend>
          <div
            class="_modal__control_42f588"
          >
            <label
              class="_modal__label_42f588"
              for="create_now"
            >
              Create Now
            </label>
            <input
              checked=""
              class="_modal__input_42f588"
              id="create_now"
              name="create_now"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_42f588"
          >
            <label
              class="_modal__label_42f588"
              for="is_active"
            >
              Is Active
            </label>
            <input
              checked=""
              class="_modal__input_42f588"
              id="is_active"
              name="is_active"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_42f588"
          >
            <label
              class="_modal__label_42f588"
              for="has_reminder_onload"
            >
              Has Reminder Onload
            </label>
            <input
              checked=""
              class="_modal__input_42f588"
              id="has_reminder_onload"
              name="has_reminder_onload"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_42f588"
          >
            <label
              class="_modal__label_42f588"
              for="has_learned_words"
            >
              Has Learned Words
            </label>
            <input
              class="_modal__input_42f588"
              id="has_learned_words"
              name="has_learned_words"
              type="checkbox"
            />
          </div>
          <div
            class="_modal__control_42f588"
          >
            <label
              class="_modal__label_42f588"
              for="sort_mode"
            >
              Sort Mode
            </label>
            <select
              class="_modal__select_42f588"
              id="sort_mode"
              name="sort_mode"
              required=""
            >
              <option
                class="_modal__option_42f588"
                value="newest"
              >
                Newest
              </option>
              <option
                class="_modal__option_42f588"
                value="oldest"
              >
                Oldest
              </option>
              <option
                class="_modal__option_42f588"
                selected=""
                value="random"
              >
                Random
              </option>
            </select>
          </div>
        </fieldset>
        <button
          class="_modal__button_42f588"
          type="submit"
        >
          Create
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
